### ISSUE TITLE
Executing repo sync commands from a third-party application

### ISSUE THREAD

**Issue by gee12** [maks/MGit#489]
Actually, the functionality is described in #448 .
If necessary, I can duplicate here.

**Comment by tlvince** [comment]
I think this is in reference to #488, not #448. 

Per [my comment](https://github.com/maks/MGit/pull/488#issuecomment-584418289) there, this could be a useful interop to enable sync for other apps.

**Comment by tlvince** [comment]
An alternative might be to add a file watcher in MGit itself; on detected change in given repos, run these commands. That way, third-party apps do not need to implement intent broadcasting and everything's contained in MGit.

**Comment by xu4wang** [comment]
> An alternative might be to add a file watcher in MGit itself; on detected change in given repos, run these commands. That way, third-party apps do not need to implement intent broadcasting and everything's contained in MGit.

Usually the scenario need a two way sync, it's hard for MGit to find a suitable time to PULL from remote repo..

**Comment by tlvince** [comment]
True, I only considered one-way sync/push (my current usecase). I currently use Syncthing for bi-directional sync, which works well, though I miss having a revision history.

**Comment by maks** [comment]
I'm not really convinced the git is suitable for use as a syncing tool. At the very least, some sort of policy is required to deal with merge conflicts. Then that policy likely needs to be user configurable and that then adds a lot of complexity.

**Comment by xeor** [comment]
Obsidian is about to have a mobile app, and a bunch of people would probably want git for syncing notes.

Currently,  gitjournal seams to be the best way.. git sync using a filewatcher would even be something I could pay for.

For the conflicts,  I think a normal notification would be sufficient

**Comment by maks** [comment]
@xeor sorry I dont know what Obsidian is. I'm not sure if using git is a good idea for these sort of "sync notes content" type of usecases, for instance what should happen if there are merge conflicts?

**Comment by xeor** [comment]
@maks could mgit just show a notification maybe, that it is unable to merge. Clicking the notification could just open the mgit app..?

**Comment by maks** [comment]
@xeor What I'm concerned about is the user knowing what to do to resolve a conflict once they are taken into MGit via the notification. Is rather not have people who are trying to use it as just a sync tool coming here or asking via Google Play comments for assistance because Git is definitely not a tool with an easy UI or even mental model for usage by developers (even some developers struggle with git concepts).

**Comment by xeor** [comment]
@maks I don't think people will try to use mgit as "just a sync tool". Most people will just use syncthing or Dropbox.

The thing is that there really isn't any good tool for this out there. Mgit is the closest and most promising.

Another alternative would be an app action (not sure what they are called,  but they can be their own icon), as a shortcut to pull/add-commit-push..? A type of widget..

Hope you are seeing some usefulness in these suggestions!

**Comment by maks** [comment]
I commented in #179  that I'm open to someone contributing this functionality to MGit via a service and that reminded me that in fact there is a [previous PR](https://github.com/maks/MGit/pull/488) that did do that, but its author never followed up on my feedback, so if anyone else wants to pickup that ball and run with it and submit a new PR I'll be ahppy to take a look at it but I don't have time at the moment to work on this directly myself.

**Comment by xeor** [comment]
Is this still an unsolved issue? :( This is one of the few ways to get git on a mobile.
* Some other tools are using an js implementation of git, which is extremely slow.
* Some others suggest using git from termux, but termux has it's own set of problems since they don't support the official playstore anymore.

**Comment by maks** [comment]
@xeor unfortunately due to restrictions from Google on apps in Google Play using the "All Files" access permission, it's likely that the usage you have in mind for MGit won't be possible unless you install a newer version of MGit from F-Droid.

**Comment by xeor** [comment]
Thanks! I tried to dig some more into why, and (for the others reading this), this is all about the security restrictions in newer android versions, like this, or making it impossible for apps to execute binaries inside it's writable directories (making embedding the git binary hard).
